src/coreclr/nativeaot/Runtime/i386/MiscStubs.asm

```diff
@@ -39,4 +40,224 @@ ProbeLoop:
+
+    cmp     ecx, 32
+    jae     LRszMORE32
+
+    ;; Handle shifts of between bits 0 and 31
+    shrd    eax, edx, cl
+    shr     edx, cl
+    ret
+
+LRszMORE32:
+    ;; Handle shifts of between bits 32 and 63
+    ;; The x86 shift instructions only use the lower 5 bits.
+    mov     eax, edx
+    xor     edx, edx
+    shr     eax, cl
+    ret
+RhpLRsz ENDP
+
+;; *********************************************************************/
+;; LMul
+;;
+;; Purpose:
+;;    Does a long multiply (same for signed/unsigned)
+;;
+;; Entry:
+;;    Parameters are passed on the stack:
+;;    1st pushed: multiplier (QWORD)
+;;    2nd pushed: multiplicand (QWORD)
+;;
+;; Exit:
+;;    EDX:EAX - product of multiplier and multiplicand
+;;
+;; NOTE: Adapted from clang generated code
+;;
+RhpLMul PROC public
+    push    esi
+    mov     ecx, dword ptr [esp + 16]
+    mov     esi, dword ptr [esp + 8]
+    mov     eax, ecx
+    mul     esi
+    imul    ecx, dword ptr [esp + 12]
+    add     edx, ecx
+    imul    esi, dword ptr [esp + 20]
+    add     edx, esi
+    pop     esi
+    ret     16
+RhpLMul ENDP
+
+;; *********************************************************************/
+;; Dbl2Lng
+;;
+;; Purpose:
+;;    Converts a double to a long truncating toward zero (C semantics)
+;;    Uses stdcall calling conventions
+;;
+;; NOTE: Adapted from JIT_Dbl2LngP4x87 in CoreCLR
+;;
+RhpDbl2Lng PROC public
+ALTERNATE_ENTRY RhpDbl2UInt
+    sub     esp, 8                  ; get some local space
+
+    fld     qword ptr [esp+0Ch]     ; fetch arg
+    fnstcw  word ptr [esp+0Ch]      ; store FPCW
+    movzx   eax, word ptr [esp+0Ch] ; zero extend - wide
+    or      ah, 0Ch                 ; turn on OE and DE flags
+    mov     dword ptr [esp], eax    ; store new FPCW bits
+    fldcw   word ptr  [esp]         ; reload FPCW with new bits
+    fistp   qword ptr [esp]         ; convert
+    mov     eax, dword ptr [esp]    ; reload FP result
+    mov     edx, dword ptr [esp+4]
+    fldcw   word ptr [esp+0Ch]      ; reload original FPCW value
+
+    add     esp, 8                  ; restore stack
+
+    ret	8
+RhpDbl2Lng ENDP
+
+;; *********************************************************************/
+;; Dbl2Int
+;;
+;; Purpose:
+;;    Converts a double to a long truncating toward zero (C semantics)
+;;    Uses stdcall calling conventions
+;;
+RhpDbl2Int PROC public
+    cvttsd2si eax, qword ptr [esp+4]
+    ret       8
+RhpDbl2Int ENDP
+
+;; *********************************************************************/
+;; Lng2Dbl
+;;
+;; Purpose:
+;;    Converts a long to a double (C semantics)
+;;    Uses stdcall calling conventions
+;;
+RhpLng2Dbl PROC public
+    fild    qword ptr [esp+4]
+    ret     8
+RhpLng2Dbl ENDP
+
+;; *********************************************************************/
+;; ULng2Dbl
+;;
+;; Purpose:
+;;    Converts an unsigned long to a double (C semantics)
+;;    Uses stdcall calling conventions
+;;
+;; NOTE: Adapted from GCC generated code
```

Comment:
This GCC mention has a high chance of getting flagged by license scanners.

This asm implementation is 3x slower compared to the C implementation that we have in CoreCLR: https://github.com/dotnet/runtime/blob/cc17166217c9e9066830c2df8374675bc2f185b3/src/coreclr/vm/jithelpers.cpp#L492

. (Same test as https://github.com/dotnet/runtime/pull/98858#issuecomment-1986379186 .)

Reply:
That’s weird. It basically does the same thing (assuming same CPU baseline). I can move it to C to unblock the PR and run some more tests locally for the perf.

Reply:
This asm implementation uses x87 floating point that is slow on any recent hardware. (Returning of the floating point value using the historic x86 calling convention seems to be special cased to be fast, everything else is slow.)

Reply:
Sure, I expect the x87 to have bad perf. I just vaguely remember this was a bit difficult with SSE2 and got easier with SSE3. I am not at my computer right now to check the details so I could very well be remembering wrong…

Reply:
> I just vaguely remember this was a bit difficult with SSE2 and got easier with SSE3.

Yes, the C-runtime implementation has the SSE3 optimization.

Reply:
I removed the ASM versions of double conversion helpers and used the existing ones in MathHelpers.cpp instead.